Task19 mybatis

andresmadeddie opened the pull request:
Create Batis configuration files, mappers, interfaces, and Dao files  for all models, and test all new features of MyBatis in the MyBatisMain

andresmadeddie committed:
create mybatis-config.xml file
create mapper files for each model class
create MyBatisSqlFactory.java
eliminate throw exception in ICategory interface

andresmadeddie committed:
update

andresmadeddie committed:
change key for Category. Update the main class to test all options in mapper

andresmadeddie committed:
change key for Category. Update the main class to test all options in mapper

andresmadeddie committed:
Person and Category Implemented 2/12

andresmadeddie committed:
Customer Implemented 3/12
create new interfaces for each class to use them in the Dao

andresmadeddie committed:
All Dao, mapper files implemented

andresmadeddie committed:
correct table names and test all MyBatis project in MyBatisMain.java file

AndresSolvd committed:
switch service classes to MyBatis

AndresSolvd committed:
switch service classes to MyBatis

AndresSolvd committed:
added files in xml and json output folder in resources

ankochergin commented on pom.xml:
```diff
@@ -87,6 +87,26 @@
+            <groupId>org.apache.cxf</groupId>
+            <artifactId>cxf-rt-rs-client</artifactId>
+            <version>4.0.1</version>
+        </dependency>
```
why you decided to use these dependencies?

ankochergin commented on src/main/java/com/solvd/MyBatisMain.java:
```diff
@@ -0,0 +1,357 @@
+        // CATEGORY
+        System.out.println("\nCategory");
+        // test Print all
+        System.out.println(new CategoryDao().getAll());
```
please use loggers instead of `System.out.println` in your app

ankochergin commented on src/main/java/com/solvd/sql/interfaces/ICustomerDao.java:
```diff
@@ -0,0 +1,6 @@
+import com.solvd.sql.model.Customer;
+
+public interface ICustomerDao extends IBaseDAO<Customer> {
+}
```
add at least one custom method to your interfaces. For example to search customer by name here

ankochergin commented on src/main/resources/mapper/orders-mapper.xml:
```diff
@@ -0,0 +1,30 @@
+<?xml version = "1.0" encoding = "UTF-8"?>
```
rename to `order-mapper.xml`

ankochergin commented on src/main/resources/mapper/productPromotion-mapper.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version = "1.0" encoding = "UTF-8"?>
```
rename to product-promotion-mapper.xml and check that other xml files also use such naming convention

ankochergin reviewed (commented):
In your models you have ids for foreign keys, try to change it to the objects. For example, instead of `customerId` add `Customer customer` object

andresmadeddie committed:
rename to order-mapper.xml, order-item-mapper.xml, product-promotion-mapper.xml. modify mybatis-config.xml accordingly.
delete cfx dependencies from pom file

andresmadeddie committed:
On the customer class, change person-id to person object, then modify "getters", "setters", "to string", a related dao file from the jdbc package, "customer-mapper.xml", and the Main Class accordingly.

andresmadeddie committed:
remove unused library

andresmadeddie committed:
OrderItem Model class
Change ids to objects, then modify "getters", "setters", "to string", Dao file from jbdc package , mapper file , and the Main Class accordingly.
remove MyBatisMain.java

andresmadeddie committed:
Orders Model class
Change ids to objects, then modify "getters", "setters", "to string", Dao file from jbdc package , mapper file , and the Main Class accordingly.

andresmadeddie committed:
For Owner, Product, ProductPromotion, Owner, Shop, Staff,   in Model class
Change ids to objects, then modify "getters", "setters", "to string", Dao file from jbdc package , mapper file , and the Main Class accordingly.

andresmadeddie committed:
Create new methods for interfaces

andresmadeddie committed:
Create new methods for interfaces an implemented 6/12

andresmadeddie committed:
implemented methods from interfaces 11/12

andresmadeddie committed:
remove unused import statement

andresmadeddie committed:
Logger implemented in Main class

ankochergin commented on src/main/java/com/solvd/sql/interfaces/IOwnerDao.java:
```diff
@@ -0,0 +1,7 @@
+import com.solvd.sql.model.Owner;
+
+public interface IOwnerDao extends IBaseDAO<Owner> {
+    Owner getByName(String name);
```
If you have the same method in each interface it should be extracted to 1 interface instead of duplicating it
Please extract it to 1 interface or add unique methods
Idea of an interface is used to declare a strict behavior that classes must implement

ankochergin commented on src/main/java/com/solvd/sql/interfaces/IOrdersDao.java:
```diff
@@ -0,0 +1,7 @@
+import com.solvd.sql.model.Orders;
+
+public interface IOrdersDao extends IBaseDAO<Orders> {
+    Orders getByCustomerName(String name);
```
nice 👍
let's just specify what we want to get in the method name: getOrderByCustomerName()

ankochergin commented on src/main/java/com/solvd/sql/interfaces/IProductPromotionDao.java:
```diff
@@ -0,0 +1,7 @@
+
+public interface IProductPromotionDao extends IBaseDAO<ProductPromotion> {
+    ProductPromotion getByProductName(String productName);
+}
```
```suggestion
public interface IProductPromotionDao extends IBaseDAO<ProductPromotion> {

    ProductPromotion getByProductName(String productName);
    
}
```
just to make it more readable

AndresSolvd committed:
modify interface methods' names and spaces

AndresSolvd committed:
add configuration file for logs setting time and date and output on a file and console

ankochergin commented on src/main/java/com/solvd/sql/jdbc/OrdersDao.java:
```diff
@@ -23,7 +23,7 @@ public void insert(Orders orders) {
         try (PreparedStatement ps = con.prepareStatement(query)) {
```
rename class to `OrderDao`

ankochergin commented on src/main/java/com/solvd/sql/model/Orders.java:
```diff
@@ -21,16 +21,16 @@ public class Orders {
     @XmlElement(name = "orderDate")
```
rename class to `Order`

ankochergin commented on src/main/java/com/solvd/sql/mybatis/OrdersDao.java:
```diff
@@ -0,0 +1,71 @@
+
+import java.util.List;
+
+public class OrdersDao implements IOrdersDao {
```
```suggestion
public class OrderDao implements IOrdersDao {
```

ankochergin commented on src/main/java/com/solvd/sql/services/OrdersService.java:
```diff
@@ -1,12 +1,12 @@
 import java.util.List;
 
-public class OrdersService implements IBaseDAO<Orders> {
+public class OrdersService implements IOrdersDao {
```
```suggestion
public class OrderService implements IOrdersDao {
```

ankochergin commented on src/main/java/com/solvd/sql/interfaces/IOrdersDao.java:
```diff
@@ -0,0 +1,8 @@
+
+import com.solvd.sql.model.Orders;
+
+public interface IOrdersDao extends IBaseDAO<Orders> {
```
```suggestion
public interface IOrderDao extends IBaseDAO<Orders> {
```

andresmadeddie committed:
rename from orders to order the model and all related implementation

jinocencio-solvd commented on src/main/java/com/solvd/sql/services/CategoryService.java:
```diff
@@ -1,7 +1,7 @@
 
 import com.solvd.sql.interfaces.ICategoryDao;
-import com.solvd.sql.jdbc.CategoryDao;
+import com.solvd.sql.mybatis.CategoryDao;
```
Instead of replacing the import statement, we can initialize the service layer object by passing the DAO through the constructor. We declare the type as the entity interface since both JDBC and MyBatis DAOs implement from a common interface.

jinocencio-solvd commented on src/main/java/com/solvd/sql/model/OrderItem.java:
```diff
@@ -12,13 +17,13 @@ public class OrderItem {
+    @XmlElement(name = "order", type = Order.class)
+    @JsonProperty("order")
+    private Order order;
 
```
Can the order have a list or collection of order items?

jinocencio-solvd commented on src/main/java/com/solvd/sql/model/Order.java:
```diff
@@ -21,16 +21,16 @@ public class Orders {
+    @XmlElement(name = "customer", type = Customer.class)
+    @JsonProperty("customer")
+    private Customer customer;
 
```
Can the order have a list or collection of order items?

andresmadeddie replied:
I will implement it on the next PR

andresmadeddie committed:
Add reset method to reset all tables at the begining of the main class

andresmadeddie committed:
Create new method that give options to create a person or to select and id from a list when the input id is returns a null object Person

andresmadeddie committed:
remove redundant initialization

andresmadeddie committed:
remove unused import statement

andresmadeddie committed:
Create new method that give options to create a owner or to select from a list of owners when the input id is returns a null object Owner

ankochergin commented on src/main/java/com/solvd/sql/services/OwnerService.java:
```diff
@@ -32,6 +45,61 @@ public List<Owner> getAll() {
+        }
+        System.out.println("New Person created: \n" + owner + "\n   *****   \n");
+        return owner;
+    }
```
interesting idea, but it's not really good here
you shouldn't have any user input or user flow here

ankochergin commented on src/main/java/com/solvd/util/SqlReset.java:
```diff
@@ -0,0 +1,49 @@
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class SqlReset {
```
```suggestion
public class SqlResetUtil {
```

ankochergin commented on src/main/java/com/solvd/util/SqlReset.java:
```diff
@@ -0,0 +1,49 @@
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
```
each statement contains `DELETE FROM ` and `ALTER TABLE ... AUTO_INCREMENT = 1;`, maybe we can optimise it somehow? probably using a loop

andresmadeddie committed:
Create creator class to manage the user's input to create persons or owners

andresmadeddie committed:
Create creator class to manage the user's input to create persons or owners

andresmadeddie committed:
Create creator class to manage the user's input to create persons or owners updates

andresmadeddie committed:
update SqlResetUtil to loop on each table to reset them

AndresSolvd committed:
Remove creator class and all its implementation in the project.
implement insert method on OwnerService to recognize if the person object in the owner object already exists in the database, if exists verify if the person already exists as an owner. Otherwise, the method will insert the person in the database add the person in the owner with the new personId and insert the new owner in the database

andresmadeddie committed:
add existence verification for Staff and Customer

andresmadeddie committed:
add existence verification for Staff and Customer

andresmadeddie committed:
update

AndresSolvd committed:
Change methods for insert in the customer, owner,and staff services

asurelaba commented on src/main/java/com/solvd/util/SqlResetUtil.java:
```diff
@@ -0,0 +1,38 @@
+            }
+
+            statement.close();
+            con.close();
```
If there is an sqlException before closing statement and con. It wont be closed or until  connection becomes stale and closed by system. It might be better to put them in finally?

andresmadeddie replied:
Agree. I improve the code

asurelaba commented on src/main/resources/mybatis-config.xml:
```diff
@@ -0,0 +1,47 @@
+        PUBLIC "-//mybatis.org//DTD Config 3.0//EN"
+        "http://mybatis.org/dtd/mybatis-3-config.dtd">
+<configuration>
+    <properties resource="db.pass"/>
```
Is this another properties file? What does it contain?

andresmadeddie replied:
It is. It is not available on the GitHub repository  and the file were the information for the connection is pulled. the db.properties is just a decoy file

AndresSolvd committed:
Fix typo in method name

AndresSolvd committed:
Identify the use of a Singleton Pattern in the connectionPool class
close statement and connection in the finally block in SqlResetUtil.java

AndresSolvd committed:
change comment word from finally to final

ankochergin commented on src/main/java/com/solvd/Main.java:
```diff
@@ -320,56 +345,158 @@ public static void main(String[] args) throws JAXBException {
+        staff = new Staff();
+        staffService = new StaffService();
+        personService = new PersonService();
+        System.out.println();
```
please remove all `System.out.println();`

AndresSolvd committed:
remove System.out.println(); from Main class/ method
